realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsDynamicTests.java

```diff
@@ -20,34 +20,484 @@
+        assertNotNull(target2Sources);
+        // TODO Is ↓ an expected behavior?
+        // if a source (in this test, source200) contains multiple references to a target in one RealmList, those must not be aggregated.
+        assertEquals(4, target2Sources.size());
+        boolean source100Found = false;
+        boolean source200Found = false;
+        for (DynamicRealmObject target2Source : target2Sources) {
+            final long idValue = target2Source.getLong(AllJavaTypes.FIELD_ID);
+            if (idValue == 100L) {
+                source100Found = true;
+            } else if (idValue == 200L) {
+                source200Found = true;
+            } else {
+                fail("unexpected id value: " + idValue);
+            }
+
+            assertEquals(AllJavaTypes.CLASS_NAME, target2Source.getType());
+            assertTrue(target2Source.getList(AllJavaTypes.FIELD_LIST).contains(target2));
+            assertFalse(target2Source.getList(AllJavaTypes.FIELD_LIST).contains(target3));
+        }
+        assertTrue(source100Found);
+        assertTrue(source200Found);
+
+        // tests sources of target3
+        final RealmResults<DynamicRealmObject> target3Sources = target3.linkingObjects(AllJavaTypes.CLASS_NAME, AllJavaTypes.FIELD_LIST);
+        assertNotNull(target3Sources);
+        assertTrue(target3Sources.isEmpty());
+
+        dynamicRealm.executeTransaction(new DynamicRealm.Transaction() {
+            @Override
+            public void execute(DynamicRealm realm) {
+                final DynamicRealmObject source200 = dynamicRealm.where(AllJavaTypes.CLASS_NAME).equalTo(AllJavaTypes.FIELD_ID, 200L).findFirst();
+                // remove last reference in the list
+                source200.getList(AllJavaTypes.FIELD_LIST).remove(2);
+            }
+        });
+
+        // backlinks are also updated
+        assertEquals(3, target2Sources.size());
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void linkingObjects_IllegalStateException_ifNotYetLoaded() {
+        final Realm realm = looperThread.realm;
+
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                final BacklinksTarget target1 = realm.createObject(BacklinksTarget.class);
+                target1.setId(1);
+
+                final BacklinksSource source = realm.createObject(BacklinksSource.class);
+                source.setChild(target1);
+            }
+        });
+
+
+        final DynamicRealm dynamicRealm = DynamicRealm.getInstance(looperThread.realmConfiguration);
+        try {
+            final DynamicRealmObject targetAsync = dynamicRealm.where(BacklinksTarget.CLASS_NAME)
+                    .equalTo(BacklinksTarget.FIELD_ID, 1L).findFirstAsync();
+            // precondition
+            assertFalse(targetAsync.isLoaded());
+
+            thrown.expect(IllegalStateException.class);
```

Comment:
I am not quite sure `thrown` works in looper thread. also, if it works, then `fail()` at line 453 is not needed.